File: src/main/scala/com/github/mauricio/async/db/postgresql/encoders/QueryMessageEncoder.scala
package com.github.mauricio.postgresql.encoders

import com.github.mauricio.postgresql.ChannelUtils
import com.github.mauricio.postgresql.messages.backend.Message
import com.github.mauricio.postgresql.messages.frontend.{FrontendMessage, QueryMessage}
import java.nio.charset.Charset
import org.jboss.netty.buffer.{ChannelBuffers, ChannelBuffer}

/**
 * User: Maurício Linhares
 * Date: 3/3/12
 * Time: 8:32 PM
 */

class QueryMessageEncoder ( charset : Charset ) extends Encoder {

  override def encode(message: FrontendMessage): ChannelBuffer = {

    val m = message.asInstanceOf[QueryMessage]

    val buffer = ChannelBuffers.dynamicBuffer()
    buffer.writeByte( Message.Query )
    buffer.writeInt(0)
    ChannelUtils.writeCString(m.query, buffer, charset)

    ChannelUtils.writeLength(  buffer )

    buffer
  }

}

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/encoder/QueryMessageEncoder.kt
package com.github.jasync.sql.db.mysql.encoder

import com.github.jasync.sql.db.mysql.message.client.ClientMessage
import com.github.jasync.sql.db.mysql.message.client.QueryMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import io.netty.buffer.ByteBuf
import java.nio.charset.Charset

class QueryMessageEncoder(val charset: Charset) : MessageEncoder {

    override fun encode(message: ClientMessage): ByteBuf {
        val m = message as QueryMessage
        val encodedQuery = m.query.toByteArray(charset)
        val buffer = ByteBufferUtils.packetBuffer(4 + 1 + encodedQuery.size)
        buffer.writeByte(ClientMessage.Query)
        buffer.writeBytes(encodedQuery)

        return buffer
    }
}
